operator/v1/types_baremetal.go

```diff
@@ -0,0 +1,68 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// Metal3ProvisioningController contains configuration used by the Provisioning service (Ironic)
+// to provision baremetal hosts.
+//
+type Metal3ProvisioningController struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the
+	// Metal3ProvisioningController.
+	Spec Metal3ProvisioningControllerSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the
+	// Metal3ProvisioningController.
+	Status Metal3ProvisioningControllerStatus `json:"status,omitempty"`
+}
+
+// ProvisioningDHCP
+type ProvisioningDHCP struct {
+	// managementState indicates whether the DHCP service would be started within
+	// the metal3 cluster or would use an external DHCP service
+	ManagementState ManagementState `json:"managementState"`
+
+	// DHCPRange
+	DHCPRange string `json:"DHCPRange,omitempty"`
+}
+
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type Metal3ProvisioningControllerSpec struct {
+	// provisioningInterface is the name of the network interface on a Baremetal server connected
+	// to the provisioning network.
+	ProvisioningInterface string `json:"provisioningInterface"`
+
+	// provisioningIP is the IP address assigned to the provisioningInterface for provisioning
+	// the baremetal node. This IP address should be within the provisioning subnet, and
+	// outside of the DHCP range.
+	ProvisioningIP string `json:"provisioningIP"`
+
+	// provisioningNetworkCIDR is the network on which the baremetal nodes are provisioned.
+	// The provisioningIP and the IPs in the dhcpRange all come from within this network.
+	ProvisioningNetworkCIDR string `json:"provisioningNetworkCIDR"`
```

Comment:
Is this informative?  What happens if not all my baremetal servers agree?

This may go back to what is the scoping of this resource: is it one per machine or one per type of machine or something else?  The IP defined above suggests its one per machine,.

Reply:
Currently it is one per cluster, we start only one metal3 deployment via the MAO, and this config data is used as input to that deployment.

That deployment is used to deploy all machines on a common provisioning subnet.

The ProvisioningIP is the IP on this subnet where the provisioning services run (those hosted by the metal3 deployment configured by this resource).  It's not one per machine, that is a cluster-wide IP.

Reply:
Can this IP change?  Is this used as input to force that IP?  It doesn't sound like it.  It sounds like this is an IP that is used to tell something else where to connect to an already existent system.  In which case this is status, not spec.

Reply:
Yes, the provisioning network already exists. But, the IP address to use during the provisioning process is not known ahead of time. So, it should be provided by the admin. 

My understanding is that fields in the Metal3ProvisioningStatus should be updated by someone/something other than the user/admin based on current state of the metal3 cluster. There is no way for metal3 to "discover" the provisioning network details. So, how can it update the Status?

Reply:
This is data provided at install time via openshift-install, which is then used by the metal3 provisioning components to configure networking appropriately.  

This will always be environment specific (so we need input from the human admin at install time), and at this time we've not considered changing any of the values as a day-2 operation, but in theory we could have that requirement in future (particularly for the DHCP Range which you might need to e.g expand after initial install).

This is *not* an IP used to tell something where to connect, it's configuration data we use to start the cluster hosted provisioning services, e.g we're starting the service using this data, which is later connected to (the baremetal controller and the provisioning services run in the same pod, started by the MAO).